nexus/db-queries/src/db/datastore/image.rs

```diff
@@ -17,50 +18,149 @@ use crate::db::error::public_error_from_diesel_pool;
 use crate::db::error::ErrorHandler;
 use crate::db::model::Image;
 use crate::db::model::Project;
+use crate::db::model::ProjectImage;
+use crate::db::model::Silo;
+use crate::db::model::SiloImage;
 use crate::db::pagination::paginated;
 
 use async_bb8_diesel::AsyncRunQueryDsl;
 
 use super::DataStore;
 
 impl DataStore {
-    pub async fn image_list(
+    pub async fn project_image_list(
         &self,
         opctx: &OpContext,
         authz_project: &authz::Project,
+        include_silo_images: bool,
         pagparams: &PaginatedBy<'_>,
     ) -> ListResultVec<Image> {
         opctx.authorize(authz::Action::ListChildren, authz_project).await?;
 
         use db::schema::image::dsl;
+        use db::schema::project_image::dsl as project_dsl;
+        match include_silo_images {
+            true => match pagparams {
+                PaginatedBy::Id(pagparams) => {
+                    paginated(dsl::image, dsl::id, &pagparams)
+                }
+                PaginatedBy::Name(pagparams) => paginated(
+                    dsl::image,
+                    dsl::name,
+                    &pagparams.map_name(|n| Name::ref_cast(n)),
+                ),
+            }
+            .filter(dsl::time_deleted.is_null())
+            .filter(
+                dsl::project_id
+                    .is_null()
+                    .or(dsl::project_id.eq(authz_project.id())),
+            )
```

Comment:
This is the special sauce of this logic branch. If an image has no `project_id` then it's a silo image.... and I'm just realizing I should probably be filtering it by silo id 😅